### ISSUE TITLE
Failed to read boolean value failure in ClickHouse connector

### ISSUE THREAD

**Issue by akenO8** [trinodb/trino#18247]
version: 417
connector: clickhouse [21.11.10.1]

query sql:
```
select
    *
from
    clickhouse.tmp_db.tmp_tb
    where stime >= timestamp '2023-07-09 00:00:00'
    and stime < timestamp '2023-07-09 05:00:00'
    and uid = 'a8f153a3'
    limit 10
```
return:
```
failed: java.io.IOException: Failed to read boolean value, expect 0 (false) or 1 (true) but we got: 58
```

This table is create in clickhouse, and the uid field is of type varchar(string).

What is the cause of this, how to solve it, thank you for your advice.

**Comment by ebyhr** [comment]
Could you share the full stacktrace? You can get it with CLI `--debug` option or Web UI.

**Comment by akenO8** [comment]
> Could you share the full stacktrace? You can get it with CLI `--debug` option or Web UI.

io.trino.spi.TrinoException: java.io.IOException: Failed to read boolean value, expect 0 (false) or 1 (true) but we got: 58
	at io.trino.plugin.jdbc.JdbcRecordCursor.handleSqlException(JdbcRecordCursor.java:314)
	at io.trino.plugin.jdbc.JdbcRecordCursor.advanceNextPosition(JdbcRecordCursor.java:188)
	at io.trino.$gen.CursorProcessor_20230712_083401_111316.process(Unknown Source)
	at io.trino.operator.ScanFilterAndProjectOperator$RecordCursorToPages.process(ScanFilterAndProjectOperator.java:335)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils$YieldingProcess.process(WorkProcessorUtils.java:182)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils$BlockingProcess.process(WorkProcessorUtils.java:208)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:318)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:360)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:347)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:241)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:256)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:146)
	at io.trino.operator.Driver.processInternal(Driver.java:402)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:305)
	at io.trino.operator.Driver.tryWithLock(Driver.java:701)
	at io.trino.operator.Driver.process(Driver.java:297)
	at io.trino.operator.Driver.processForDuration(Driver.java:268)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:888)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:561)
	at io.trino.$gen.Trino_417____20230703_030724_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.sql.SQLException: java.io.IOException: Failed to read boolean value, expect 0 (false) or 1 (true) but we got: 58
	at com.clickhouse.jdbc.SqlExceptionUtils.create(SqlExceptionUtils.java:45)
	at com.clickhouse.jdbc.SqlExceptionUtils.handle(SqlExceptionUtils.java:90)
	at com.clickhouse.jdbc.ClickHouseResultSet.next(ClickHouseResultSet.java:716)
	at io.trino.plugin.jdbc.JdbcRecordCursor.advanceNextPosition(JdbcRecordCursor.java:185)
	... 33 more
Caused by: java.io.IOException: Failed to read boolean value, expect 0 (false) or 1 (true) but we got: 58
	at com.clickhouse.data.ClickHouseInputStream.readBoolean(ClickHouseInputStream.java:937)
	at com.clickhouse.data.format.BinaryDataProcessor$NullableDeserializer.deserialize(BinaryDataProcessor.java:120)
	at com.clickhouse.data.format.ClickHouseRowBinaryProcessor.readAndFill(ClickHouseRowBinaryProcessor.java:260)
	at com.clickhouse.data.ClickHouseDataProcessor.nextRecord(ClickHouseDataProcessor.java:134)
	at com.clickhouse.data.ClickHouseDataProcessor.access$000(ClickHouseDataProcessor.java:25)
	at com.clickhouse.data.ClickHouseDataProcessor$RecordsIterator.next(ClickHouseDataProcessor.java:51)
	at com.clickhouse.data.ClickHouseDataProcessor$RecordsIterator.next(ClickHouseDataProcessor.java:37)
	at com.clickhouse.jdbc.ClickHouseResultSet.next(ClickHouseResultSet.java:714)
	... 34 more

**Comment by ebyhr** [comment]
Thanks! Please also share:
* `CREATE TABLE` & `INSERT` statements in ClickHouse (not Trino)
* clickhouse.properties in Trino

**Comment by akenO8** [comment]
> tmp_db.tmp_tb

CREATE TABLE:
`CREATE TABLE IF NOT EXISTS tmp_db_lo.tmp_tb ON CLUSTER ck (
    `stime`                    DateTime                    comment 'stime',
    `uid`                       Nullable(String)            comment 'uid'
)
ENGINE=ReplicatedMergeTree(
    '/clickhouse/tables/{layer}-{shard}/tmp_db_lo/tmp_tb',
    '{replica}' 
)
ORDER BY (stime)
SETTINGS index_granularity = 8192;

CREATE TABLE IF NOT EXISTS tmp_db.tmp_tb ON CLUSTER ck (
    `stime`                    DateTime                    comment 'stime',
    `uid`                       Nullable(String)            comment 'uid'
)
ENGINE=Distributed(ck,tmp_db_lo,tmp_tb)`

INSERT:
I imported the csv file data through clickhouse-client.

clickhouse.properties:
`        "connector.name=clickhouse", 
        "connection-url=jdbc:clickhouse://clickhouse:post", 
        "connection-user=trino", 
        "connection-password=passwd", 
        "case-insensitive-name-matching=true", 
        "clickhouse.map-string-as-varchar=true", 
        "insert.non-transactional-insert.enabled=true"`

I found a couple of special uid values in this data(A string full of numbers), and I don't know if they're relevant.
`select
    uid
from
    ck.tmp_db.tmp_tb
where stime >= timestamp '2023-07-09 00:00:00'
    and stime < timestamp '2023-07-09 05:00:00'
group by uid
order by uid
limit 10`
`000
00000
000000000000
000000000003
...`

**Comment by ebyhr** [comment]
Please use three backtick ` ``` ` for code snippet.

**Comment by ebyhr** [comment]
> I imported the csv file data through clickhouse-client.

Please share the reproducible steps. We can't help with the provided information.

**Comment by akenO8** [comment]
> > I imported the csv file data through clickhouse-client.
> 
> Please share the reproducible steps. We can't help with the provided information.

1. The impala data is converted into a csv file.
2. Use clickhouse-client to insert tb.
`
clickhouse-client -m -n -t --compression=1 --host ck-lb.internal --port 9111 --user 'ck' --password xxx --input_format_allow_errors_num 0 --input_format_allow_errors_ratio 0 --insert_deduplicate 0 --date_time_input_format 'best_effort' --format_csv_delimiter "" --query "insert into tmp_db_lo.tmp_tb format CSV"  < /tmp/20230709/s000000001
`

**Comment by ebyhr** [comment]
Also, can you select the table in ClickHouse side without Trino?

**Comment by akenO8** [comment]
> a8f153a3

yes i can

**Comment by akenO8** [comment]
> Also, can you select the table in ClickHouse side without Trino?

yes i can

**Comment by ebyhr** [comment]
> yes i can , the same query

What client did you use? I would recommend debugging with JDBC client because the connector uses JDBC driver internally. If it's a bug of JDBC driver, you should file an issue in https://github.com/ClickHouse/clickhouse-java

**Comment by akenO8** [comment]
I used superset to connect to clickhouse via JDBC, it is a JDBC client.

**Comment by ebyhr** [comment]
https://superset.apache.org/docs/databases/clickhouse/ mentions Python client though.

If you really use ClickHouse JDBC driver with Superset, I would recommend finding the difference, e.g. driver version. 
Also, please note that Trino converts SELECT statement to ClickHouse syntax. If you executed SELECT statement of the issue description, it's different from the internal query of the connector.

**Comment by akenO8** [comment]
Thanks for your reminder, I found the execution history of this sql pushdown to clickhouse.
```
DB::Exception: Estimated query execution time (390.4502161118316 seconds) is too long. Maximum: 180.
```
Then I adjust `timeout_before_checking_execution_speed` to 60 to avoid this problem.


And then I had a question, why is this sql executing so fast in clickhouse, but it is so slow when trino to connect to clickhouse.
And I noticed that `limit` didn't push down to clickhouse.

On the performance of this sql on both sides:
clickhouse: 8s
trino: 80s
